Validate track id as object path and other improvements

Tomin1 opened the pull request:
Main change here is allowing that `QDBusObjectPath` use with `QVariant` and validate that a valid path can be constructed from the passed track id value. The other changes are adding logging categories for better, easier and more granular logging enablement and that small fix to initialization of some string variables.

Validate that `trackId` is a valid object path before setting it. If it happens to be `QDBusObjectPath` already, then convert it to string properly. Null `QVariant` is passed through as it becomes NoTrack path once converted to `QDBusObjectPath` when accessed over D-Bus. Invalid values print a warning and set `trackId` to null.

This metadata C++ API is overall a bit strange. It takes QVariant objects but expects certain (undocumented!) types or otherwise things don't work as expected. I think having typed accessors would be much more ergonomic, but I understand that this was built QML first.

cc: @pvuorela

Tomin1 committed:
mpris: Add logging categories

[mpris] Add logging categories

Signed-off-by: Tomi Leppänen <tomi.leppanen@jolla.com>

Tomin1 committed:
mpris: Fix invalid initialization in MprisPlayerInterface

[mpris] Fix invalid initialization in MprisPlayerInterface

Using QStringLiteral for most efficient construction. Since these values
are not passed outside the library it should be safe even if the library
was unloaded.

Signed-off-by: Tomi Leppänen <tomi.leppanen@jolla.com>

Tomin1 committed:
Improve gitignore

Signed-off-by: Tomi Leppänen <tomi.leppanen@jolla.com>

abranson commented on src/mprisplayerinterface.cpp:
```diff
@@ -40,10 +41,10 @@ MprisPlayerInterface::MprisPlayerInterface(const QString &service, const QString
     , m_canPlay(false)
     , m_canSeek(false)
-    , m_loopStatus(Mpris::LoopNone)
+    , m_loopStatus(QMetaEnum::fromType<Mpris::LoopStatus>().valueToKey(Mpris::LoopNone))
```
Might this be better the other way around, leaving this the same and doing conversion on setting and getting instead?

Tomin1 replied:
Maybe, but it seems this wouldn't be exactly a trivial change as the D-Bus interface still must look outside like it has strings. These are stored as part of the interface class and their setter and getter use methods from the parent class.

abranson replied:
That's not very nice. String constants instead that are already the right value?

Tomin1 replied:
Do you mean that I should have `m_loopStatus("LoopNone")` instead of using `QMetaEnum`?

I tried a bit if I could replace the underlying implementation with enum values but at least the first try seems to just lead to a crash. I think I should make a task about it and leave it for later as there are more important things to figure out right now.

Tomin1 commented on src/mprismetadata.cpp:
```diff
@@ -272,9 +275,15 @@ namespace {
 
-        if (from.isNull()) {
+        if (from.isNull() || !canConvert) {
+            if (!canConvert)
```
This should be `!from.isNull() && !canConvert` instead of just `!canConvert`.

Tomin1 committed:
mpris: Register object path before service

[mpris] Register object path before service

This avoids errors caused by unfortunate timing:
mpris-proxy: No such object path '/org/mpris/MediaPlayer2'

Signed-off-by: Tomi Leppänen <tomi.leppanen@jolla.com>

Tomin1 commented:
~~Note that I simplified things quite a lot by ensuring that metadata always contains track id as `QDBusObjectPath` but `MprisMetaData::trackId()` will return `QString` in `QVariant`. As stated in the relevant commit message:~~

> Simplified the code by always storing mpris:trackid as object path,
> which is what mpris specification expects, and then always returning
> QString from MprisMetaData::trackId().

~~This allows to make some parts of `MprisClient` simpler as it doesn't need to deal with both types.~~

Edit: Dropped this. It would possibly break situations where this was already working to some extent.

Tomin1 commented on src/mprismetadata.cpp:
```diff
@@ -417,8 +426,9 @@ MprisMetaData::MprisMetaData(QObject *parent)
 
 QVariant MprisMetaData::trackId() const
 {
+    // Returns trackId in string format although for D-Bus we want it to look like an object path
```
Oops. :bug:
Later it doesn't return `trackId` value but `priv->m_metaData[MetaFieldTrackId]` directly.

Tomin1 replied:
Fixed.

abranson commented on src/mprismetadata.cpp:
```diff
@@ -417,20 +426,22 @@ MprisMetaData::MprisMetaData(QObject *parent)
     if (priv->m_metaData.contains(MetaFieldTrackId)) {
-        QString trackId = priv->m_metaData[MetaFieldTrackId].toString();
+        QString trackId = priv->m_metaData[MetaFieldTrackId].value<QDBusObjectPath>().path();
         if (trackId.startsWith(TrackObjectPathPrefix)) {
```
I think this bit should be removed while you're here. I don't think we should be prepending anything really. It's up to the player to come up with a sane trackID

Tomin1 replied:
I was going to wonder if that might break something but this code was never used because the line before was just wrong. Perhaps removing it would be the best thing to do.

Forcing this to be QDBusObjectPath or QString probably breaks more than removing these lines.

abranson replied:
Yeah although it's a separate issue, it's about the only piece of this method not touched so it seems silly to preserve it. And you're right, it's never done anything, but could cause problems if it decides to. Probably best to remove the inverse operation at the same time though.

Tomin1 replied:
> I was going to wonder if that might break something but this code was never used because the line before was just wrong. Perhaps removing it would be the best thing to do.

Except... it would have worked if the value stored was `QString` which it likely had been since `QDBusObjectPath` would not have survived to be delivered over D-Bus (conversion happens in `typedMetaData()` which I missed apparently). I think I need to revert my thinking a bit and maybe go back to converting the values in `MprisClient` when fetching them, otherwise I break those ways that this actually could have worked at all.

I should probably first take a small break though to clear my thoughts so I won't spend time uselessly going back and forth any more. Sorry for all the hassle.

Tomin1 replied:
I'm not going to remove that now after all.

abranson reviewed (approved):
Removing that prefix stuff is minor really. This is fine with or without it.

Tomin1 commented:
I went back to the version that just ensures that we are correctly converting QDBusObjectPath instances to strings using them as such. Now it shouldn't introduce any regressions either for existing users.

Tomin1 committed:
mpris: Validate trackId as object path

[mpris] Validate trackId as object path. JB#59304

Validate that trackId is a valid object path before setting it. If it
happens to be QDBusObjectPath already, then convert it to string
properly. Null QVariant is passed through as it becomes NoTrack path
once converted to QDBusObjectPath when accessed over D-Bus. Invalid
values print a warning and set trackId to null.

Signed-off-by: Tomi Leppänen <tomi.leppanen@jolla.com>

Tomin1 commented:
Made this convert QDBusObjectPath instances to strings when storing them and added validation for track id value if it something else than that or null. I think this simplified it a lot.

abranson commented on src/mprismetadata.cpp:
```diff
@@ -430,7 +439,21 @@ QVariant MprisMetaData::trackId() const
+        // Pass, null QVariant makes D-Bus interface to advertise NoTrack value
+    } else {
+        QString id = trackId.toString();
+        if (!validateTrackIdAsObjectPath(id))
```
Is this necessary as a separate function?

Tomin1 replied:
No, it's not, but I thought it's more readable this way.

Tomin1 replied:
I know I can squash all the validation on one line of code but then it looks messy. The name of the function should give at least a little bit of idea what is happening there and hopefully the compiler is smart enough to inline it.
